### PR TITLE
Proposal: Prevent orphan volumes creation from missing rm -v

### LINKED ISSUE #8363: PROPOSAL: docker volumes list command
### Problem
Volumes remain a hidden piece within Docker.

1. Weather you are using normal volumes or host-mounted volumes, there is no way to get information about these volumes except by inspecting a container to find where these volumes are and manually inspecting them.

2. There tend to be orphaned volumes because the only way to remove a volume _correctly_ is by doing `docker rm -v <container>`, but most people do not think to do this.

3. The most basic usage of a volume is to store state beyond the container's own lifecycle, however if you remove a container, it is mostly impossible to use that volume again.  Because of this, we've invented the concept of a "data-only" container so it can hold a reference to that volume for you and you can just reference the container instead.  This is pretty heavy on the management side and creates container clutter, for every volume there is a corresponding container that does nothing but hold a volume reference.

### Proposal
New command: `docker volumes ls`, which produces output as such:
```
NAME                   CREATED                USED COUNT
jolly_torvalds         42 seconds ago         2
romatic_wozniak        About a minute ago     1
```
Optionally include the volumes size with `docker volumes ls --size`

Volumes would be automatically assigned a name when created through "docker run -v"
Through a separate command a volume could be created outside of the `docker run` and a name could be specified by the user or generated if none was specified (just as with containers).

New command: `docker volumes inspect <voume>`, which produces output as such:
```json
{
        "ID": "124670d406cb783e37ad10d339e7f980db75460bcf4fedf1baa00566ebbcad63",
        "Name": "jolly_torvalds",
        "Created": "2014-10-01T16:17:01.417634973Z",
        "Path": "/var/lib/docker/vfs/dir/124670d406cb783e37ad10d339e7f980db75460bcf4fedf1baa00566ebbcad63",
        "IsBindMount": false,
        "Writable": true,
        "Containers": [
             "04b22da1ac4c47f758a3553d5c1c36e7510d1e99c80da19504f5f9112bc5491e",
             "835f0b9b62d09145d14339ee2f651df5d8dec1239c3c94b2cdc8a614649ecc1e"
         ],
        "Aliases": [
             "determined_brown:/foo",
             "angry_bell:/bar"
        ],
        "Size": "38325000"
}
```
Aliases are defined as the `container_name:/mount/path` for the container using a given volume.

### Using an existing volume
Volumes can be referenced by `docker run` as such: `docker run -v jolly_torvalds:/foo`

Volumes can also be referenced using `--volumes-from` such as: `docker run --volumes-from determined_brown:/foo`, which is a volume alias.

### Naming addendum
Volume's names will use the same random name generator as containers, but in a separate namespace.  This means that a volume and a container, which may be completely unrelated to each other, could have the exact same name.
This also means a volume's alias (which would be a consuming container's name + the mount path in the container) may include a name that is also attached to a volume.

### PR DESCRIPTION
**Pull request by allingeek**
Added logic to delete.go preventing the removal of containers that would create orphan managed volumes.

If volumes are not being removed then check that none of the target container's volumes are only referenced by the target container. If such a volume exists return an appropriate error and stop the container removal.

Fixes #8363 - By fixing the orphan issue, not by creating list-volumes

Signed-off-by: Jeff Nickoloff jeff@allingeek.com

### CODE REVIEW

File: daemon/delete.go [hunk 1]
```diff
@@ -61,6 +61,17 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
 				return job.Errorf("Conflict, You cannot remove a running container. Stop the container before attempting removal or use -f")
 			}
 		}
+                // require volume removal if the container has the sole reference to a volume
+                if !removeVolume {
+                        for id := range container.VolumePaths() {
+                                if v := daemon.volumes.Get(id); v != nil && !v.IsBindMount {
+                                        containers := v.Containers()
+                                        if len(containers) == 1 {
+                                                return job.Errorf("Cannot remove a container with sole ownership of volumes. Volume %s must be removed ith -v.", v.Path)
```
Comment (thaJeztah) [inline daemon/delete.go]: Typo ith -> with

**Comment by thaJeztah** [comment]
While I agree that "orphaned" containers can be a nuisance, I don't think this is the proper way to handle this. Not automatically deleting a volume when deleting a container is (afaik) *by design*. It allows removing a container without permanently destroying the data generated by that container.

Also, software like fig/docker-compose rely on this when re-creating containers; they destroy the old container (*keeping* the volume), then create a new container and attach the volume of the old container.

Docker surely needs some improvement to handle "orphaned" volumes, because right now, they are basically invisible, unless you know where to look for them.

There's a proposal that handles this, here: https://github.com/docker/docker/pull/8484. While that PR has not yet been accepted, the author of that PR also created a utility that does the same, but as a separate tool; you can find it here: https://github.com/cpuguy83/docker-volumes

**Comment by icecrime** [comment]
Ping @cpuguy83: thoughts?

**Comment by cpuguy83** [comment]
I'd probably change the error message, but I like this.
It helps make sure volumes are cleaned up without just blindly deleting the underlying data.

**Comment by allingeek** [comment]
It may be that fig/compose would need to migrate to using the same volume-container pattern that everyone else uses. But the plain use-case should not be ignored. Using a separate utility or running some maintenance script to cleanup is technical debt. It would be better to fix the problematic behavior directly. This is especially problematic when users are unaware that volumes have been created by a container and potentially filled with data. It is a side effect that leaves a machine in a bad state.

**Comment by cpuguy83** [comment]
I don't think this is much different than not being allowed to remove a dir that is not empty in the shell (via `rm`, vs `rm -r`), this is why I like this solution.

**Comment by thaJeztah** [comment]
@allingeek agreed (wrt fig/compose), there are also plans to change the "recreate" step in compose; https://github.com/docker/compose/issues/874

I'm still a bit hesitant; I can imagine cases where you'd want to delete a container, but preserve the volume. But yes, you'd probably be able to do so by adding an dummy container using `--volumes-from`, or exporting the volume.

**Comment by cpuguy83** [comment]
I'm changing this to code review because I think this design looks good (and sense there's been no movement from anyone else).

@allingeek this one needs a rebase if you're still up for it.
Thanks!

**Comment by allingeek** [comment]
I'll take care of that this morning.

**Comment by LK4D4** [comment]
This needs to be very carefully documented. Also I'm still not sure about this. It'll break 99% scripts, because `-v` option is pretty unknown to community.

**Comment by allingeek** [comment]
I agree that this should be well documented in the release notes. Making people aware of the issue that their scripts are creating is sort of the point. I haven't seen much evidence that -v is unknown though. This is the only way I'm aware of to remove a Docker managed volume without manual intervention.

**Comment by thaJeztah** [comment]
> I haven't seen much evidence that -v is unknown though.

Well, there are lots of reports of "Docker leaking disk space"; many of those turn out to be people that aren't aware that volumes won't be removed when doing `docker rm` (or `docker run --rm ...`).

**Comment by cpuguy83** [comment]
All the more reason:)

**Comment by thaJeztah** [comment]
> All the more reason:)

Agree with you there, but it *is* a breaking change. Still hoping the volumes-management proposal will make its way back somehow, which will make "dangling" volumes no longer a problem (because they can be managed even without containers attached to them)

**Comment by allingeek** [comment]
Relatively speaking, this is a very polite breaking change. It improves the platform and prevents tools built around it from leaking resources. It also provides a hint for a three character change that adopters can make to fix their integration. 

I like volume management as well, but adopting that sort of thing would be a much more significant undertaking. A major augmentation to the UX is maybe a bit drastic just to avoid dealing with a funky interface problem. People will forgive the short term pain.

**Comment by allingeek** [comment]
Looks like I'll need to update the integration tests.

**Comment by allingeek** [comment]
The current integration test that is failing tests container removal with the -f option. The help text for -f reads, "Force the removal of a running container (uses SIGKILL)" I'm going to modify the integration test rather than adjust the functionality because this option is specifically about stopping a running container.

**Comment by cpuguy83** [comment]
Moving this back to design-review, I think I was being a bit too pushy, need another maintainer to ok design... sorry @allingeek

**Comment by stevvooe** [comment]
This needs a closer look from @icecrime and @crosbymichael.

If I understand this correctly, a breaking change is being introduced to expose a possible resource leakage. If there is no other mechanism to expose this to a user (ie list-volumes from #8363), then this would be considered an acceptable tradeoff.

**Comment by LK4D4** [comment]
ping @tiborvass @jfrazelle @icecrime @crosbymichael 
I'm okay with this change, but we need to decide.

**Comment by tiborvass** [comment]
@allingeek @cpuguy83 I'm torn between *this is cool for today* and *this is not what we want long term*. I would much rather have volume management like in @cpuguy83's proposal. You should definitely want to be able to remove a container without removing the volume. The problem is that today they are too much linked.

Should we accept this PR because it solves a problem today, knowing that we would want to revert the behavior in the future? Maybe.

**Comment by cpuguy83** [comment]
I like that it forces someone people to do something with their data.
In cases where a new container is replacing the existing one they can just go ahead and rename the current one, do volumes-from, then delete the old one.
If they truly don't need the data anymore, then do the `rm -v`.

**Comment by allingeek** [comment]
I'd love volume management, but until that is real this lack of visibility
of the problem will continue to be a risk for adopters.

**Comment by allingeek** [comment]
I don't think that first class volumes fix the UX problem. The real issue here is visibility on orphan volumes. First-class volume operations help with volume management, but do nothing for orphan visibility at container deletion time. Even with the features proposed in #8363, a user should still be presented with a warning when they delete the only container attached to a volume.

**Comment by tiborvass** [comment]
@allingeek except that if you had volumes as first class, you could list orphaned ones and delete them.

**Comment by allingeek** [comment]
Yes, you could list them. But unless you are informed that you've created one, or you have regular asynchronous pruning as part of your disciplined workflow, you will never be prompted to perform a cleanup.

**Comment by tiborvass** [comment]
@allingeek what about the usecase where you do want orphaned volumes? I guess you could ask for the user to create via the volumes commands instead of creating it when creating a container. But then it's an asymmetry... Idk :S

**Comment by cpuguy83** [comment]
Yes, I would prefer the prevention of garbage vs forcing someone to clean it up manually.
Also, we don't have a volumes command yet :)
Last I heard we wanted that to be out of process anyway.

**Comment by thaJeztah** [comment]
>Yes, I would prefer the prevention of garbage vs forcing someone to clean it up manually.

Well, from a UX perspective, at least it's consistent with exited containers because users didn't run with `--rm` and "dangling" images. Whether that's good or bad, that's a different thing :-)

**Comment by allingeek** [comment]
There are two UX issues. First, it is not possible to cleanup orphan volumes with docker. Second, volumes that can be created as invisible side-effects can be orphaned without warning to the user. The proposed change both prevents the creation of orphaned volumes and informs the user. Existing tools that do not depend on the orphan volume behavior can migrate with a two character change. Those that do depend on that behavior can use the same volume-container pattern that everyone else uses.

Maintaining problematic behavior because other parts of the project exhibit similar, but wholly unrelated problematic behavior is not a positive UX trait. I could never imagine a user thinking to themselves, "I'm sure glad Docker didn't merge a fix to this disk leak. It would have really confused me."

**Comment by cpuguy83** [comment]
ping @docker/core-maintainers Can we get a final decision on this?

**Comment by duglin** [comment]
First, we need some testcases.
Second, given a choice between today's design (creating orphans w/o knowing it and not being able to clean them up) and this, I'd prefer this PR.

**Comment by cpuguy83** [comment]
Well, it's in design-review, so let's not ask to add new code when the design is not approved.

**Comment by duglin** [comment]
oops - always forget to check that  :-)

**Comment by duglin** [comment]
ping @crosbymichael since its a UX change.

**Comment by LK4D4** [comment]
ping @docker/core-maintainers

**Comment by calavera** [comment]
This PR as become totally obsolete, none of the methods used here exist on Docker master anymore.

I'm going to close it for now. This will become easier to implement and understand from an UX point of view after a few more improvements in volume management.

**Comment by thaJeztah** [comment]
For those following this; top level volume management is now implemented in https://github.com/docker/docker/pull/14242, which was just merged, and will be in **docker 1.9**

If you want to test it before that, nightly builds, build from master can be found at https://master.dockerproject.org

**Pull request closed without merging**